feat: threads (#692)
discordjs/guide: guide/popular-topics/threads.md

- `Client#threadListSync`: Emitted whenever the client user gains access to a text or news channel that contains threads.
- `Client#threadMembersUpdate`: Emitted whenever members are added or removed from a thread. Requires `GUILD_MEMBERS` privileged intent.
- `Client#threadMemberUpdate`: Emitted whenever the client user's thread member is updated.

## Creating and deleting threads

Threads are created and deleted using the `ThreadManager` of a text or news channel.
To create a thread you call the `ThreadManager#create()` method:

```js {2-8}
client.on('messageCreate', async message => {
	const thread = await message.channel.threads
		.create({
			name: 'food-talk',
			autoArchiveDuration: 60,
			reason: 'Needed a separate thread for food',
		});
	console.log(`Created thread: ${thread.name}`);
});
```

To delete a thread you call the `ThreadChannel#delete()` method:

```js {2,3}
client.on('messageCreate', async message => {
	const thread = message.channel.threads.cache.find(x => x.name === 'food-talk');
	await thread.delete();
});
```

## Joining and leaving threads

> To join your Client to a ThreadChannel you make use of the `ThreadChannel#join()` method:

Reviewer: Either here or in the private vs public section something about checking whether the thread is joinable maybe? (really only matters for private)

```

And to leave one you call `ThreadChannel#leave()`;

```js {3}
client.on('messageCreate', async message => {
	const thread = message.channel.threads.cache.find(x => x.name === 'food-talk');
	await thread.leave();
});
```

## Archiving, unarchiving and locking threads

Every thread can be either active or archived. Changing a thread from archived to active is referred to as unarchiving the thread. Threads that have `locked` set to true can only be unarchived by a member with the `MANAGE_THREADS` permission.

Threads automatically archive after inactivity. 'Activity' is defined as sending a message, unarchiving a thread, or changing the auto-archive time.

To archive a thread you use the `ThreadChannel#setArchived()` method and pass `true` as parameter:

```js {3}
client.on('messageCreate', async message => {
	const thread = message.channel.threads.cache.find(x => x.name === 'food-talk');
	await thread.setArchived(true);
});
```

And to unarchive it one you pass `false ` to one and the same method:

```js {3}
client.on('messageCreate', async message => {
	const thread = message.channel.threads.cache.find(x => x.name === 'food-talk');
	await thread.setArchived(false);
});
```

> This same principle applies to locking and unlocking a thread via the `ThreadChannel#setLocked()` method:

Reviewer: Probably some note about not being able to lock when the thread is archived would be useful

# Threads

Threads can be thought of as temporary sub-channels inside an existing channel, to help better organize conversation in a busy channel.

## Thread related gateway events

::: tip
You can use the `isThread()` typeguard to make sure the channel you're working with is a ThreadChannel!
:::

Threads introduce a number of new gateway events, which are listed below:

- `Client#threadCreate`: Emitted whenever a thread is created or when the client user is added to a thread.
- `Client#threadDelete`: Emitted whenever a thread is deleted.
- `Client#threadUpdate`: Emitted whenever a thread is updated - e.g. name change, archive state change, locked state change.
- `Client#threadListSync`: Emitted whenever the client user gains access to a text or news channel that contains threads.
- `Client#threadMembersUpdate`: Emitted whenever members are added or removed from a thread. Requires `GUILD_MEMBERS` privileged intent.
- `Client#threadMemberUpdate`: Emitted whenever the client user's thread member is updated.

## Creating and deleting threads

> Threads are created and deleted using the `ThreadManager` of a text or news channel.
> To create a thread you call the `ThreadManager#create()` method:
> 
> ```js {2-8}
> client.on('messageCreate', async message => {
> 	const thread = await message.channel.threads
> 		.create({
> 			name: 'food-talk',
> 			autoArchiveDuration: 60,
> 			reason: 'Needed a separate thread for food',
> 		});
> 	console.log(`Created thread: ${thread.name}`);
> });

Reviewer 2: ```suggestion
Threads are created and deleted using the `ThreadManager` of a text or news channel.
To create a thread you call the `ThreadManager#create()` method:

<!-- eslint-skip -->

```js
const thread = await message.channel.threads.create({
	name: 'food-talk',
	autoArchiveDuration: 60,
	reason: 'Needed a separate thread for food',
});

console.log(`Created thread: ${thread.name}`);
```

Reviewer 2: Removing the `client.on('messageCreate', ...)`  wrapper since it doesn't entirely make sense. Unfortunately it means adding an `<!-- eslint-skip -->` comment at each codeblock because of top level await usage, but they're all small snippets so it's no big deal.

```

> To delete a thread you call the `ThreadChannel#delete()` method:
> 
> ```js {2,3}
> client.on('messageCreate', async message => {
> 	const thread = message.channel.threads.cache.find(x => x.name === 'food-talk');
> 	await thread.delete();
> });

Reviewer 2: ```suggestion
To delete a thread, use the `ThreadChannel#delete()` method:

```js
const thread = message.channel.threads.cache.find(x => x.name === 'food-talk');
await thread.delete();
```

```

## Joining and leaving threads

> To join your Client to a ThreadChannel you make use of the `ThreadChannel#join()` method:
> 
> ```js {3}
> client.on('messageCreate', async message => {
> 	const thread = message.channel.threads.cache.find(x => x.name === 'food-talk');
> 	if (thread.joinable) await thread.join();
> });

Reviewer 2: ```suggestion
To join your client to a ThreadChannel, use the `ThreadChannel#join()` method:

<!-- eslint-skip -->

```js
const thread = message.channel.threads.cache.find(x => x.name === 'food-talk');
if (thread.joinable) await thread.join();
```

```

> And to leave one you call `ThreadChannel#leave()`;
> 
> ```js {3}
> client.on('messageCreate', async message => {
> 	const thread = message.channel.threads.cache.find(x => x.name === 'food-talk');
> 	await thread.leave();
> });

Reviewer 2: ```suggestion
And to leave one, use `ThreadChannel#leave()`;

<!-- eslint-skip -->

```js
const thread = message.channel.threads.cache.find(x => x.name === 'food-talk');
await thread.leave();
```

```

> ## Archiving, unarchiving and locking threads

Reviewer 2: ```suggestion
## Archiving, unarchiving, and locking threads
```

> Every thread can be either active or archived. Changing a thread from archived to active is referred to as unarchiving the thread. Threads that have `locked` set to true can only be unarchived by a member with the `MANAGE_THREADS` permission.
> 
> Threads automatically archive after inactivity. 'Activity' is defined as sending a message, unarchiving a thread, or changing the auto-archive time.

Reviewer 2: ```suggestion
A thread can be either active or archived. Changing a thread from archived to active is referred to as unarchiving the thread. Threads that have `locked` set to true can only be unarchived by a member with the `MANAGE_THREADS` permission.

Threads are automatically archived after inactivity. "Activity" is defined as sending a message, unarchiving a thread, or changing the auto-archive time.
```

> To archive a thread you use the `ThreadChannel#setArchived()` method and pass `true` as parameter:
> 
> ```js {3}
> client.on('messageCreate', async message => {
> 	const thread = message.channel.threads.cache.find(x => x.name === 'food-talk');
> 	await thread.setArchived(true);
> });

Reviewer 2: ```suggestion
To archive or unarchive a thread, use the `ThreadChannel#setArchived()` method and pass in a boolean parameter:

<!-- eslint-skip -->

```js
const thread = message.channel.threads.cache.find(x => x.name === 'food-talk');
await thread.setArchived(true); // archived
await thread.setArchived(false); // unarchived
```

```

> And to unarchive it one you pass `false ` to one and the same method:
> 
> ```js {3}
> client.on('messageCreate', async message => {
> 	const thread = message.channel.threads.cache.find(x => x.name === 'food-talk');
> 	await thread.setArchived(false);
> });
> ```

Reviewer 2: ```suggestion
```

> This same principle applies to locking and unlocking a thread via the `ThreadChannel#setLocked()` method:
> 
> Locking a thread:
> 
> ```js {3}
> client.on('messageCreate', async message => {
> 	const thread = message.channel.threads.cache.find(x => x.name === 'food-talk');
> 	await thread.setLocked(true);
> });

Reviewer 2: ```suggestion
This same principle applies to locking and unlocking a thread via the `ThreadChannel#setLocked()` method:

<!-- eslint-skip -->

```js 
const thread = message.channel.threads.cache.find(x => x.name === 'food-talk');
await thread.setLocked(true); // locked
await thread.setLocked(false); // unlocked
```

```

::: warning
Archived threads can't be locked!
:::

> Unlocking a thread:
> 
> ```js {3}
> client.on('messageCreate', async message => {
> 	const thread = message.channel.threads.cache.find(x => x.name === 'food-talk');
> 	await thread.setLocked(false);
> });
> ```
> 

Reviewer 2: ```suggestion
```

## Public and private threads

> Public threads are viewable by everyone who can view the parent channel of the thread. Public threads must be created from an existing message, but can be 'orphaned' if that message is deleted. The created thread and the message it was started from will share the same id. The type of thread created matches the type of the parent channel.

Reviewer 2: ```suggestion
Public threads are viewable by everyone who can view the parent channel of the thread. Public threads must be created from an existing message, but can be "orphaned" if that message is deleted. The created thread and the message it originated from will share the same ID. The type of thread created matches the parent channel's type.
```

Private threads behave similar to Group DMs, but in a Guild. Private threads can only be created on text channels.

> To create a private thread you use `ThreadManager#create()` as above, but pass `private_thread` as `type`, like so:
> 
> ```js {6}
> client.on('messageCreate', async message => {
> 	const thread = await message.channel.threads
> 		.create({
> 			name: 'mod-talk',
> 			autoArchiveDuration: 60,
> 			type: 'private_thread',
> 			reason: 'Needed a separate thread for moderation',
> 		});
> 	console.log(`Created thread: ${thread.name}`);
> });

Reviewer 2: ```suggestion
To create a private thread, use `ThreadManager#create()` and pass in `private_thread` as the `type`:

```js {4}
const thread = await message.channel.threads.create({
	name: 'mod-talk',
	autoArchiveDuration: 60,
	type: 'private_thread',
	reason: 'Needed a separate thread for moderation',
});

console.log(`Created thread: ${thread.name}`);
```

```

## Adding and removing members

> You can add and remove members to and from a thread channel.
> 
> To add a member to a thread you can use the `ThreadMemberManager#add()` method like shown below:
> 
> ```js {3}
> client.on('messageCreate', async message => {
> 	const thread = message.channel.threads.cache.find(x => x.name === 'food-talk');
> 	await thread.members.add('140214425276776449');
> });

Reviewer 2: ```suggestion
You can add and remove members to and from a thread channel.

To add a member to a thread, use the `ThreadMemberManager#add()` method:

<!-- eslint-skip -->

```js
const thread = message.channel.threads.cache.find(x => x.name === 'food-talk');
await thread.members.add('140214425276776449');
```

```

> And to remove a member from a thread, you repeat this procedure using `ThreadMemberManager#remove()`:
> 
> ```js {3}
> client.on('messageCreate', async message => {
> 	const thread = message.channel.threads.cache.find(x => x.name === 'food-talk');
> 	await thread.members.remove('140214425276776449');
> });

Reviewer 2: ```suggestion
And to remove a member from a thread, use `ThreadMemberManager#remove()`:

<!-- eslint-skip -->

```js
const thread = message.channel.threads.cache.find(x => x.name === 'food-talk');
await thread.members.remove('140214425276776449');
```
